woocommerce/woocommerce-ios
# [Orders with Coupons M7] Product discount components

Closes: #10837

## Description
This PR adds the updated "Discount View"  to a product, which involves:
- Updating the design of the existing `CollapsibleProductRowCard` to show an "Add discount" button as well as the discount after has been added.
- Adding a new `ProductDiscountView` and `DiscountLineDetailsView`, where the discount calculation and application is performed. This substitutes the previous modal view. 

As with the collapsable product row, the design is not 1:1 with the [proposed design](https://www.figma.com/file/FTOWxedKUyDbBdOrxRfMyI/Coupons-in-Order-Creation?type=design&node-id=14-10245&mode=design&t=Tn0eG6fym2kCjRo7-0), but gets as close as possible. Items I left outside

<img width="275" alt="Screenshot 2023-10-09 at 16 04 09" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/6f44bbc6-66b2-4c8a-88b1-0e4d82256a6e">

- There is no percentage discount calculation in this view. We just show the discount and the total discounted amount.
- The unique button to switch between fixed and percentage is two buttons for simplicity.

## Caveats:
 - We kept the changes to `FeeOrDiscountLineDetailsViewModel` to the minimum, since this part of the project conflicts with Custom Amounts ( pdfdoF-3Sp-p2 cc: @toupper ). Ideally we'll want to circle back and decouple entirely fees from discounts.
 - Along with the previous point, there are known bugs with the existing discounts implementation (for example being able to enter positive discounts, or being able to discount more than the total price of the product. We're not addressing these with this PR, but are logged in https://github.com/woocommerce/woocommerce-ios/issues/10828 and https://github.com/woocommerce/woocommerce-ios/issues/5143)

![Simulator Screen Recording - iPhone 15 - 2023-10-09 at 15 48 41](https://github.com/woocommerce/woocommerce-ios/assets/3812076/1138d02a-9454-410b-ac67-0231287c8579)

## Testing instructions
* Switch `ordersWithCouponsM6` feature flag to true
* Go to `Orders` > `+` > `+ Add Products` > Add a product to your order, expand the chevron.
* Observe that:
  * ` +Add discount` button is available
  * Tapping on it opens the new product discount view
  * We can enter either a fixed or a percentage amount of discount. Once we do so, the calculation happens on-screen. Note that here is the bug where we can add a discount bigger than the total price of the product. This also renders a `Remove Discount` button.
  * Tapping `Add` adds the discount to the product and dismisses the view. 
  * Now we can see a `Discount` + pencil icon button as well as the discount added to the product. Tapping again allows to edit or remove the discount.

| 1 | 2 | 3 | 4 |
|--------|--------|--------|--------|
| <img width="295" alt="Screenshot 2023-10-09 at 16 12 46" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/2f47dca7-04e5-41a0-a53d-de37a4898350"> | <img width="292" alt="Screenshot 2023-10-09 at 16 13 10" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/f3c7b9a8-cea7-48ea-a339-dc98cd451689"> | <img width="293" alt="Screenshot 2023-10-09 at 16 13 29" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/5f153533-bebf-4e34-8099-c164164f4a4d"> | <img width="296" alt="Screenshot 2023-10-09 at 16 13 49" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/1e594582-f36b-44f9-b279-3a71f9c482d1"> |
